homeassistant/core.py

```diff
@@ -247,62 +324,110 @@ def __eq__(self, other):
+    def async_listeners(self):
+        """Dict with events and the number of listeners.
+
+        This method must be run in the event loop.
+        """
+        return {key: len(self._listeners[key])
+                for key in self._listeners}
 
     @property
     def listeners(self):
         """Dict with events and the number of listeners."""
-        with self._lock:
-            return {key: len(self._listeners[key])
-                    for key in self._listeners}
+        return run_callback_threadsafe(
+            self._loop, self.async_listeners
+        ).result()
 
     def fire(self, event_type: str, event_data=None, origin=EventOrigin.local):
         """Fire an event."""
         if not self._pool.running:
             raise HomeAssistantError('Home Assistant has shut down.')
 
-        with self._lock:
-            # Copy the list of the current listeners because some listeners
-            # remove themselves as a listener while being executed which
-            # causes the iterator to be confused.
-            get = self._listeners.get
-            listeners = get(MATCH_ALL, []) + get(event_type, [])
+        self._loop.call_soon_threadsafe(self.async_fire, event_type,
+                                        event_data, origin)
+        return
 
-            event = Event(event_type, event_data, origin)
+    def async_fire(self, event_type: str, event_data=None,
+                   origin=EventOrigin.local, wait=False):
+        """Fire an event.
 
-            if event_type != EVENT_TIME_CHANGED:
-                _LOGGER.info("Bus:Handling %s", event)
+        This method must be run in the event loop.
+        """
+        # Copy the list of the current listeners because some listeners
+        # remove themselves as a listener while being executed which
+        # causes the iterator to be confused.
+        get = self._listeners.get
+        listeners = get(MATCH_ALL, []) + get(event_type, [])
 
-            if not listeners:
-                return
+        event = Event(event_type, event_data, origin)
+
+        if event_type != EVENT_TIME_CHANGED:
+            _LOGGER.info("Bus:Handling %s", event)
+
+        if not listeners:
+            return
 
-            job_priority = JobPriority.from_event_type(event_type)
+        job_priority = JobPriority.from_event_type(event_type)
 
-            for func in listeners:
-                self._pool.add_job(job_priority, (func, event))
+        sync_jobs = []
+        for func in listeners:
+            if asyncio.iscoroutinefunction(func):
+                self._loop.create_task(func(event))
+            else:
+                sync_jobs.append((job_priority, (func, event)))
+
+        # Send all the sync jobs at once, since there's a lock involved we
+        # fire this off to a thread as well
+        if sync_jobs:
+            try:
+                self._pool.add_many_jobs(sync_jobs)
+            except:
+                _LOGGER.info("Unable to queue: %s", sync_jobs)
```

Comment:
Should be `_LOGGER.warning`.

Actually, what could be a reason that this will raise?